src/wav.rs

```diff
@@ -0,0 +1,388 @@
+
+            chunk.write(&mut writer)?;
+
+            // Update the riff chunk size:
+            riff_chunk.size = riff_chunk
+                .size
+                .checked_add(CHUNK_HEADER_LEN)
+                .ok_or(Error::new(ErrorKind::InvalidInput, "RIFF max size reached"))?;
+
+            chunk
+        };
+
+        // Write the tag:
+
+        tag.write_to(&mut writer, version)?;
+
+        id3_chunk.size = writer
+            .stream_position()?
+            .try_into()
+            .expect("ID3 chunk max size reached");
+
+        // Add padding if necessary.
+        if id3_chunk.size % 2 == 1 {
+            let padding = [0];
+            writer.write_all(&padding)?;
+            id3_chunk.size = id3_chunk
+                .size
+                .checked_add(padding.len() as u32)
+                .ok_or(Error::new(
+                    ErrorKind::InvalidInput,
+                    "ID3 chunk max size reached",
+                ))?;
+        }
+
+        // We must flush manually to prevent silecing write errors.
+        writer.flush()?;
+    }
+
+    // Update chunk sizes in the file:
+
+    file.seek(id3_chunk_pos)?;
+    id3_chunk.write(&file)?;
+
+    riff_chunk.size = riff_chunk
+        .size
+        .checked_add(id3_chunk.size)
+        .ok_or(Error::new(ErrorKind::InvalidInput, "RIFF max size reached"))?;
+
+    file.seek(riff_chunk_pos)?;
+    riff_chunk.write(&file)?;
+
+    Ok(())
+}
+
+/// Locates the RIFF and ID3 chunks, returning their headers. The ID3 chunk may not be
+/// present. Returns a pair of (RIFF header, ID3 header).
+fn locate_relevant_chunks(
+    mut input: impl Read + Seek,
+) -> crate::Result<(ChunkHeader, Option<ChunkHeader>)> {
+    // We must scope this reader to prevent conflict with the following writer.
+    let mut reader = BufReader::new(&mut input);
+
+    let riff_chunk = ChunkHeader::read_riff_header(&mut reader)?;
+
+    // Prevent reading past the RIFF chunk, as there may be non-standard trailing data.
+    let eof = riff_chunk
+        .size
+        .checked_sub(WAVE_TAG.len()) // We must disconsider the WAVE tag that was already read.
+        .ok_or(Error::new(
+            ErrorKind::InvalidInput,
+            "Invalid RIFF chunk size",
+        ))?;
+
+    let id3_chunk = match ChunkHeader::find_id3(&mut reader, eof.into()) {
+        Ok(chunk) => Some(chunk),
+        Err(Error {
+            kind: ErrorKind::NoTag,
+            ..
+        }) => None,
+        Err(error) => return Err(error),
+    };
+
+    // BufReader may read past the id3 chunk. To seek back to the chunk, we must first
+    // drop the BufReader, and then seek.
+    let pos = reader.stream_position()?;
+    drop(reader);
+    input.seek(SeekFrom::Start(pos))?;
+
+    Ok((riff_chunk, id3_chunk))
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+struct ChunkTag(pub [u8; 4]);
+
+impl ChunkTag {
+    /// Checks if two chunk tags match, case insensitive.
+    pub fn is(&self, tag: &ChunkTag) -> bool {
```

Comment:
I made it a separate function because it uses a case insensitive comparison. The tags may not be strictly equal, but they may represent the same entity. For instance, we can have a file with a "id3 " chunk, which will be compared to the const "ID3 " tag in the program. Are you sure it's a good idea to have a case insensitive Eq instance?

Reply:
ChunkTag is a representation of the same concept within the file format. If the format specifies that it is case insensitive then deriving Eq will result in the wrong implementation. Also, having two comparison operators is confusing :)

Reply:
Okay, I'll do it.